### PR TITLE
(feat) Implement the inline date picker with showDate and shownDateOptions

### PR DESCRIPTION
**Pull request by hadijahkyampeire**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary
- This PR implements the conditional appending of an inline date picker to a question when `showDate: true` is defined in the parent question. The value of this date field is then submitted as the `obsDatetime` for that question.
- The PR also handles some validations on the date picker which are passed to the `showDateOptions` object.
- Example

```
{
  "name": "Hadijah Test form",
  "pages": [
    {
      "label": "page 1",
      "sections": [
        {
          "label": "Section one",
          "isExpanded": "true",
          "questions": [
            {
              "label": "Height",
              "type": "obs",
              "required": false,
              "id": "height",
              "questionOptions": {
                "rendering": "number",
                "concept": "5090AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
                "showDate": "true",
                "showDateOptions": {
                  "validators": [
                    {
                      "type": "date",
                      "allowFutureDates": "false"
                    }
                  ]
                },
                "conceptMappings": [
                  {
                    "relationship": "SAME-AS",
                    "type": "AMPATH",
                    "value": "5090"
                  },
                  {
                    "relationship": "SAME-AS",
                    "type": "CIEL",
                    "value": "5090"
                  },
                  {
                    "relationship": "SAME-AS",
                    "type": "PIH",
                    "value": "5090"
                  }
                ],
                "disallowDecimals": true
              },
              "validators": []
            },
            {
              "label": "Weight",
              "type": "obs",
              "required": false,
              "id": "weight",
              "questionOptions": {
                "rendering": "number",
                "showDate": "true",
                "showDateOptions": {
                  "validators": [
                    {
                      "type": "date",
                      "allowFutureDates": "false"
                    }
                  ],
                  
                  "hide": {
                    "hideWhenExpression": "isEmpty(weight)"
                  }
                  
                },
                "concept": "5089AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
                "conceptMappings": [
                  {
                    "relationship": "SAME-AS",
                    "type": "PIH",
                    "value": "5089"
                  },
                  {
                    "relationship": "SAME-AS",
                    "type": "AMPATH",
                    "value": "5089"
                  },
                  {
                    "relationship": "SAME-AS",
                    "type": "CIEL",
                    "value": "5089"
                  }
                ],
                "disallowDecimals": true
              },
              "validators": []
            }
          ]
        }
      ]
    }
  ],
  "processor": "EncounterFormProcessor",
  "encounterType": "ca3aed11-1aa4-42a1-b85c-8332fc8001fc",
  "referencedForms": [],
  "uuid": "f6ad56a8-9f5b-465c-92be-e4ba7368f838",
  "description": "Inline Date test",
  "version": "1.0"
}
```

## Screenshots

https://github.com/openmrs/openmrs-form-engine-lib/assets/30952856/538d73d4-9186-4f64-a355-2a7c27c4eefe

## Related Issue
- https://openmrs.atlassian.net/browse/O3-3052

### CODE REVIEW

**Review by samuelmale — COMMENTED** [review]
Thanks for the work @hadijahkyampeire!

I'm concerned about how we currently manage expressions vs booleans in general within the schema. Eg. 

1. Hide

We two props:

```ts
interface FormField {
  hide?: HideProps;
  isHidden?: boolean;
}

export interface HideProps {
  hideWhenExpression: string;
}

```

2. Required

```ts
interface FormField {
  isRequired?: boolean;
  required?: string | boolean | RequiredFieldProps;
}

interface RequiredFieldProps {
  type: string;
  message?: string;
  referenceQuestionId: string;
  referenceQuestionAnswers: Array<string>;
}
```

3. Readonly 

For the readonly, the engine assumes this to either be an expression (string) or a boolean, so it will attempt to evaluate it directly, [see](https://github.com/openmrs/openmrs-form-engine-lib/blob/main/src/components/encounter/encounter-form.component.tsx#L253-L266). 

I need your thoughts on:

1. The ideal approach for the disable prop. (What is the best place to store the expression)
2. How can we harmonise the above to have a more consistent pattern on how to manage expressions in general?

cc: @ibacher @eudson @denniskigen

File: src/components/encounter/encounter-form.component.tsx [hunk 1]
```diff
@@ -378,6 +386,31 @@ const EncounterForm: React.FC<EncounterFormProps> = ({
...
   };
 
+  // disable expression
+  const evalDisable = (node, allFields: FormField[], allValues: Record<string, any>) => {
+    const { value, type } = node;
+    const isDisabled = evaluateExpression(value['disable']?.disableWhenExpression, node, allFields, allValues, {
+      mode: sessionMode,
+      patient,
+    });
+    node.value.disabled = isDisabled;
```
[the remark this answers is not in the document]
Comment (samuelmale) [reply src/components/encounter/encounter-form.component.tsx:396]: After a second look, I think I was confused by `disabled` vs `disable` and I ended up worrying about overwriting the expression; an example of how ambiguous this is.

File: src/types.ts [hunk 2]
```diff
@@ -216,6 +217,7 @@ export interface FormQuestionOptions {
   maxLength?: string;
   minLength?: string;
   showDate?: string;
+  showDateOptions?: { validators?: Array<Record<string, any>>, hide?: { hideWhenExpression: string; }, disable?: DisableProps }
```
[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply src/types.ts:220]: No actually, I don't think so, I will check

[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply src/types.ts:220]: Actually @samuelmale I think we should leave the disable in the `showDateOptions`, I have not seen the `showDate `object but I have seen the disable object somewhere, https://github.com/openmrs/openmrs-ngx-formentry/blob/8180b25f26df261276e8eb92269b1324be4d13bd/projects/ngx-formentry/src/triage.json#L139

[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply src/types.ts:220]: The word is actually called `shownDateOptions` , I was missing the `n`. check [this](https://github.com/openmrs/openmrs-ngx-formentry/blob/8180b25f26df261276e8eb92269b1324be4d13bd/src/app/adult-1.4.json#L3812)

File: src/types.ts [hunk 3]
```diff
@@ -127,6 +127,7 @@ export interface FormField {
   historicalExpression?: string;
   constrainMaxWidth?: boolean;
   meta?: QuestionMetaProps;
+  disable?: DisableProps;
```
[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply src/types.ts:130]: Actually, that line you shared is what I added and it is an object with a `disableWhenExpression` string.

[the remark this answers is not in the document]
Comment (samuelmale) [reply src/types.ts:130]: @hadijahkyampeire that what pseudo code, and yes we need to support expressions specific to the disable property of a field. My point is we have to be systematic and consistent with how we deal with the schema.

File: src/components/encounter/encounter-form.component.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (hadijahkyampeire) [reply src/components/encounter/encounter-form.component.tsx:396]: @samuelmale the `disable` object is like the hide, I thought `required` is just a boolean

**Comment by hadijahkyampeire** [comment]
@samuelmale I think the PR is ready to go in, please take another look.

**Comment by hadijahkyampeire** [comment]
@samuelmale moving the evalutions to the formHelpers puts them in a circular import and that's why I have moved them to the expression-runner, check the failing tests.

**Comment by samuelmale** [comment]
Yes @hadijahkyampeire I'm looking into it.

**Review by samuelmale — APPROVED** [review]
Nice work @hadijahkyampeire

**Comment by hadijahkyampeire** [comment]
Thanks @samuelmale

**Pull request merged**